ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs

```diff
@@ -348,8 +357,14 @@ bracketChainSyncClient
         withWatcher
             "ChainSync.Client.rejectInvalidBlocks"
             (invalidBlockWatcher varState)
-      $ LeakyBucket.execAgainstBucket lopBucketConfig
-      $ \lopBucket ->
+      $ LeakyBucket.execAgainstBucket'
+      $ \lopBucket -> do
+            -- NOTE: If `updateLopBucketConfig` was in STM, then all the
+            -- following could be in the same STM transaction which would make
+            -- it significantly more robust.
+            initialGsmState <- atomically getGsmState
+            updateLopBucketConfig lopBucket initialGsmState
+            atomically $ modifyTVar varHandles $ Map.update (\csch -> Just $! csch{cschGsmCallback = updateLopBucketConfig lopBucket}) peer
```

Comment:
I think it would be worth to get this right; races in this area seem like a horrifying surprise in the future. AFAICT, the non-STM stuff is: 

 1. `getMonotonicTime` at the very beginning.

    This can probably be solved by sth like `m (STM m ())` as we talked about previously (the `m` takes the time and returns an `STM` transaction that captured it). This relies on the fact that the transaction will complete promptly, but that is not new, the current implementation of `snapshotFill` already relies on that.
 2. Potentially stopping/starting the background thread.

    Maybe this could be done by an action *returned* by the `STM` transaction?

Reply:
Wouldn't the best solution here to give up on pure STM endpoints (because that will just not happen and I'm not sure how to make sure we won't have races with the thread starting and stopping, for instance). We could use another kind of semaphore, though, like `putMVar` before all this and `takeMVar` after all this?

Reply:
Yeah, using non-STM synchronization also sounds fine, anything that makes sure that there are no races if the GSM state changes at almost the same time we are registering a new ChainSync peer.

Reply:
Fix in https://github.com/IntersectMBO/ouroboros-consensus/pull/1007/commits/dc6ca25347f2984df61e495056e3983764aa1b23 à la https://github.com/IntersectMBO/ouroboros-consensus/pull/1007#issuecomment-2070080993.